### PR TITLE
[RFR] Refactor <ReferenceInput>

### LINKED ISSUE #4531: Migrate useReferenceInputController to use useDataProvider instead of sagas
The requests made by the `<ReferenceInput>` component aren't cached because their controller still uses the `CrudGetMatchingAccumulate` action creator. That means that the `dataProvider.getList()` it makes to fetch the possible references is done by the old fetch saga (called by the `accumulate` saga) instead of the `useDataProvider` hook. As the cache is only implemented in the new hook, `<ReferenceInput>` doesn't use the cache.

We need to migrate the `useReferenceInputController` code so that it doesn't use the accumulate saga anymore.

Seen in #4386

### LINKED ISSUE #4601: Empty view shows after coming back from detail view
1. Open the url `/#/posts/1` on the simple example on  a new window
2. Click on the Save button
3. The user is redirected to the List view, which shows the Empty component... but it shouldn't.

Tested in v3.3.3

### PR DESCRIPTION
**Pull request by fzaninotto**
## Problem

The `<ReferenceInput>` code lags behind the rest of the code, it is hard to follow, and it only allows limited functionality in children. 

## Solution

- [x] Have `<ReferenceInput>` implement a `<ListContext>` (refs #4992)
- [x] Implement deduplication logic from accumulate saga into `useQueryWithStore`
- [x] Make `useReferenceInput` use `useDataProvider` instead of sagas (Closes #4531)
- [x] Fix List view shows Empty component when coming from the Edit view (closes #4601)
- [ ] ~~Move loading logic from `<ReferenceInput>` to child (to be consistent with `<List>`)~~ (even ReferenceField doesn't do that yet)

### CODE REVIEW

File: packages/ra-core/src/controller/input/useReferenceInputController.ts [hunk 1]
```diff
@@ -77,61 +60,195 @@ interface Option {
...
+        },
+        dataStatus: {
+            error: dataStatus.error,
+            loading: dataStatus.waiting,
+            warning: dataStatus.warning,
+        },
+        choices: Object.keys(finalData).map(id => finalData[id]),
+        // kept for backwards compatibility
+        // @deprecated to be removed in 4.0
         error: dataStatus.error,
```
Comment (Luwangel) [inline packages/ra-core/src/controller/input/useReferenceInputController.ts:200]: Only the `error` is deprecated or it's also the following parameters?

File: packages/ra-core/src/controller/useListController.ts [hunk 2]
```diff
@@ -177,7 +177,7 @@ const useListController = <RecordType extends Record = Record>(
         get(state.admin.resources, [resource, 'list', 'ids'], [])
     );
     const defaultTotal = useSelector((state: ReduxState): number =>
-        get(state.admin.resources, [resource, 'list', 'total'], 0)
+        get(state.admin.resources, [resource, 'list', 'total'])
```
Comment (Luwangel) [inline packages/ra-core/src/controller/useListController.ts:180]: If you don't pass 0, the result could be "undefined". It could lead to errors if the sub-components don't manage this use case.

**Review by Luwangel — CHANGES_REQUESTED** [review]
Top! I only have a few remarks and questions.

File: packages/ra-ui-materialui/src/list/pagination/Pagination.tsx [hunk 3]
```diff
@@ -75,7 +75,7 @@ const Pagination: FC<PaginationProps> = props => {
     );
 
     // Avoid rendering TablePagination if "page" value is invalid
-    if (total === 0 || page < 1 || page > totalPages) {
+    if (total === null || total === 0 || page < 1 || page > totalPages) {
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/list/pagination/Pagination.tsx:78]: fortunately this is JS, where `null/25=0` with no warning.

File: packages/ra-core/src/controller/input/useReferenceInputController.ts (hunk 1 above)
Comment (fzaninotto) [reply packages/ra-core/src/controller/input/useReferenceInputController.ts:200]: it's the following parameters, but we'll need the RecordContext to truly deprecate them

File: packages/ra-core/src/controller/useListController.ts (hunk 2 above)
Comment (fzaninotto) [reply packages/ra-core/src/controller/useListController.ts:180]: Normally they all do. That's why I fixed the Pagination component. 

As for user land components, yes this is a BC break, albeit minor in my opinion. And it's necessary to fix #4601

**Comment by fzaninotto** [comment]
Switching to RFR

**Comment by fzaninotto** [comment]
I'm getting a wrning when updating a Comment in the simple example, back to WIP

**Comment by fzaninotto** [comment]
And back to RFR

File: packages/ra-core/src/controller/input/index.ts [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import ReferenceArrayInputController from './ReferenceArrayInputController';
 import ReferenceInputController from './ReferenceInputController';
-import useReferenceInputController from './useReferenceInputController';
+import useReferenceInputController, {
```
Comment (djhi) [inline packages/ra-core/src/controller/input/index.ts:3]: Probably a good opportunity to use only named exports

**Review by djhi — APPROVED** [review]

**Review by djhi — APPROVED** [review]
